File: shipbob_to_gorgias/csharp/sb2gorgias/Json.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Sb2Gorgias;

/// <summary>
/// The serializer options both API clients share, and accessors for the JSON this
/// integration does not model.
/// </summary>
public static class Json
{
    /// <summary>
    /// Web defaults skip the many fields both APIs send that this integration does not
    /// read; the naming policy maps their snake_case onto the PascalCase records in
    /// <c>ShipBob.cs</c>.
    /// </summary>
    public static readonly JsonSerializerOptions Options = new(JsonSerializerDefaults.Web)
    {
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
    };

    /// <summary>Deserialize <paramref name="text"/> using <see cref="Options"/>.</summary>
    public static T? Deserialize<T>(string text) => JsonSerializer.Deserialize<T>(text, Options);

    // Accessors for untyped JSON. Each takes a nullable receiver and answers null for a
    // missing key, a JSON null, or a value of the wrong shape, so a path such as
    // `order.Field("recipient").Field("address").AsText()` reads straight through with no
    // null checks along the way. JsonNode's own indexer would throw on the last two.

    /// <summary>The value at <paramref name="key"/>, or null unless this is an object with that key.</summary>
    public static JsonNode? Field(this JsonNode? node, string key) =>
        node is JsonObject item && item.TryGetPropertyValue(key, out var value) ? value : null;

    /// <summary>The value as a string, or null if it is absent or is not a JSON string.</summary>
    public static string? AsText(this JsonNode? node) =>
        node is JsonValue value && value.TryGetValue<string>(out var text) ? text : null;

    /// <summary>The value as a whole number, or null if it is absent or is not one.</summary>
    public static long? AsNumber(this JsonNode? node) =>
        node is JsonValue value && value.TryGetValue<long>(out var number) ? number : null;

    /// <summary>The value as an object, or null if it is absent or is not one.</summary>
    public static JsonObject? AsMap(this JsonNode? node) => node as JsonObject;

    /// <summary>The value's elements, or nothing at all if it is absent or is not an array.</summary>
    public static IReadOnlyList<JsonNode?> AsList(this JsonNode? node) =>
        node is JsonArray array ? [.. array] : [];
}

File: shipbob_to_gorgias/kotlin/app/src/main/kotlin/sb2gorgias/Json.kt
@file:OptIn(ExperimentalSerializationApi::class)

package sb2gorgias

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNamingStrategy
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.longOrNull

/**
 * The JSON codec both clients share. These settings are what let the ShipBob shapes in
 * `ShipBob.kt` stay small.
 */
val json: Json = Json {
    // Both APIs send far more than this integration reads, and add fields over time.
    ignoreUnknownKeys = true
    // ShipBob uses a missing key and an explicit null interchangeably — an OnHold
    // shipment sends `"tracking": null` where another topic simply omits the key.
    // Coercing turns both into the property's declared default.
    coerceInputValues = true
    // Both APIs speak snake_case; the data classes stay camelCase, as Kotlin does.
    namingStrategy = JsonNamingStrategy.SnakeCase
}

// --- reading JSON this integration does not model -----------------------------
//
// These accessors all take a *nullable* receiver, so reading into a raw order needs no
// `?.` and no intermediate checks: `order["recipient"]["address"]["city"].string`. A
// missing key, a JSON `null`, and a value of the wrong shape all answer `null`.

/** The value at [key], or `null` unless this really is an object with that key. */
operator fun JsonElement?.get(key: String): JsonElement? = (this as? JsonObject)?.get(key)

/** The value as a string, or `null` if it is absent or is not a JSON string. */
val JsonElement?.string: String?
    get() = (this as? JsonPrimitive)?.takeIf { it.isString }?.content

/** The value as a whole number, or `null` if it is absent or is not one. */
val JsonElement?.long: Long?
    get() = (this as? JsonPrimitive)?.longOrNull

/** The value as an object, or `null` if it is absent or is not one. */
val JsonElement?.obj: JsonObject?
    get() = this as? JsonObject

/** The value's elements, or nothing at all if it is absent or is not an array. */
val JsonElement?.list: List<JsonElement>
    get() = (this as? JsonArray).orEmpty()
